fuel/utils.py

```diff
@@ -14,6 +16,347 @@
+        else:
+            # If the indexable is an HDF5 dataset, it only supports fancy
+            # indexing with sorted lists. As a workaround, if
+            # `safe_hdf5_indexing` is set to `True`, Subset will do the
+            # indexing in sorted order, and reshuffle the result in the
+            # original order.
+            if isinstance(indexable, h5py.Dataset) and safe_hdf5_indexing:
+                return self.safe_unsorted_fancy_index(indexable, request)
+            # If
+            #   a) the indexable is a numpy array, or
+            #   b) it's an HDF5 dataset and the request is a sorted list of
+            #      indices,
+            # then the request can be processed directly.
+            elif isinstance(indexable, (numpy.ndarray, h5py.Dataset)):
+                return indexable[request]
+            # Anything else (e.g. lists) isn't considered to support fancy
+            # indexing, so Subset does it manually.
+            else:
+                return iterable_fancy_indexing(indexable, request)
+
+    def _is_list(self, list_or_slice):
+        """Determines if an object is a list or a slice.
+
+        Parameters
+        ----------
+        list_or_slice : :class:`list` or :class:`slice`
+            It is assumed to be one or the other, **and nothing else**.
+
+        Returns
+        -------
+        rval : bool
+            `True` if the object is a list, `False` if it's a slice.
+
+        """
+        return not hasattr(list_or_slice, 'step')
+
+    @property
+    def is_list(self):
+        """Whether this subset is list-based (as opposed to slice-based)."""
+        return self._is_list(self.list_or_slice)
+
+    @property
+    def num_examples(self):
+        """The number of examples this subset spans."""
+        if self.is_list:
+            return len(self.list_or_slice)
+        else:
+            start, stop, step = self.slice_to_numerical_args(
+                self.list_or_slice, self.original_num_examples)
+            # The problem of finding the number of examples with start > 0
+            # reduces to the problem of finding the number of examples with
+            # start' = 0 and stop' = stop - start (assuming start < stop, which
+            # is enforced in _slice_subset_sanity_check).
+            stop = stop - start
+            start = 0
+            # We count the number of times (stop - 1) is divisible by step
+            # (because stop is defined exclusively), plus 1 (because we count
+            # the zero index).
+            return (stop - 1) // step + 1
+
+    def _subset_sanity_check(self, list_or_slice, num_examples):
+        if self._is_list(list_or_slice):
+            self._list_subset_sanity_check(list_or_slice, num_examples)
+        else:
+            self._slice_subset_sanity_check(list_or_slice, num_examples)
+
+    def _list_subset_sanity_check(self, indices, num_examples):
+        if len(indices) == 0:
```

Comment:
But on the other hand people might create an empty subset intentionally. It
might be useful as a neutral element in the   subset algebra.

On 18 September 2015 at 10:13, vdumoulin <notifications@github.com> wrote:

> In fuel/utils.py
> <https://github.com/mila-udem/fuel/pull/223#discussion_r39858510>:
>
> > +            # is enforced in _slice_subset_sanity_check).
> > +            stop = stop - start
> > +            start = 0
> > +            # We count the number of times (stop - 1) is divisible by step
> > +            # (because stop is defined exclusively), plus 1 (because we count
> > +            # the zero index).
> > +            return (stop - 1) // step + 1
> > +
> > +    def _subset_sanity_check(self, list_or_slice, num_examples):
> > +        if self._is_list(list_or_slice):
> > +            self._list_subset_sanity_check(list_or_slice, num_examples)
> > +        else:
> > +            self._slice_subset_sanity_check(list_or_slice, num_examples)
> > +
> > +    def _list_subset_sanity_check(self, indices, num_examples):
> > +        if len(indices) == 0:
>
> All requests on that subset would return empty arrays, I'm not sure
> there's a use case for that. Plus, bugs caused by accidentally creating an
> empty subset may be hard to debug.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mila-udem/fuel/pull/223/files#r39858510>.
>